pkg/networkservice/chains/client/client.go

```diff
@@ -74,19 +80,35 @@ func NewClient(ctx context.Context, name string, onHeal *networkservice.NetworkS
 	return rv
 }
 
-// NewCrossConnectClientFactory - returns a (2.) case func(cc grpc.ClientConnInterface) NSM client factory.
-func NewCrossConnectClientFactory(name string, onHeal *networkservice.NetworkServiceClient, tokenGenerator token.GeneratorFunc, additionalFunctionality ...networkservice.NetworkServiceClient) func(ctx context.Context, cc grpc.ClientConnInterface) networkservice.NetworkServiceClient {
+// Generator is a type for networkservice.NetworkServiceClient generator function
+type Generator func(ctx context.Context, cc grpc.ClientConnInterface) networkservice.NetworkServiceClient
+
+// NewClientFactory - returns a (2.), (3.) cases func(cc grpc.ClientConnInterface) NSM client factory.
+func NewClientFactory(
+	name string,
+	onHeal *networkservice.NetworkServiceClient,
+	tokenGenerator token.GeneratorFunc,
+	additionalFunctionalityGenerators ...Generator,
+) func(ctx context.Context, cc grpc.ClientConnInterface) networkservice.NetworkServiceClient {
 	return func(ctx context.Context, cc grpc.ClientConnInterface) networkservice.NetworkServiceClient {
-		return chain.NewNetworkServiceClient(
-			mechanismtranslation.NewClient(),
-			NewClient(ctx, name, onHeal, tokenGenerator, cc, additionalFunctionality...),
-		)
+		var additionalFunctionality []networkservice.NetworkServiceClient
+		for _, additionalFunctionalityFactory := range additionalFunctionalityGenerators {
+			additionalFunctionality = append(additionalFunctionality, additionalFunctionalityFactory(ctx, cc))
+		}
+		return NewClient(ctx, name, onHeal, tokenGenerator, cc, additionalFunctionality...)
 	}
 }
 
-// NewClientFactory - returns a (3.) case func(cc grpc.ClientConnInterface) NSM client factory.
-func NewClientFactory(name string, onHeal *networkservice.NetworkServiceClient, tokenGenerator token.GeneratorFunc, additionalFunctionality ...networkservice.NetworkServiceClient) func(ctx context.Context, cc grpc.ClientConnInterface) networkservice.NetworkServiceClient {
-	return func(ctx context.Context, cc grpc.ClientConnInterface) networkservice.NetworkServiceClient {
-		return NewClient(ctx, name, onHeal, tokenGenerator, cc, additionalFunctionality...)
+// FromClient is a common networkservice.NetworkServiceClient wrapper to Generator
+func FromClient(client networkservice.NetworkServiceClient) Generator {
```

Comment:
Do we really need this? I guess we should not use already created clients.

Reply:
We can have some stateless client chain elements, so there is no need to create a new instance for them for each client.
E.g.: 
https://github.com/networkservicemesh/sdk/blob/1f553739f0c327c154994322a4f9f5e09dce1763/pkg/networkservice/chains/nsmgr/server_test.go#L328

Reply:
you're correct, but I think it is a complex way. I feel we should not consider is a chain element stateless or not.

Reply:
OK, fixed.